src/mesh_smoothing.rs

```diff
@@ -91,6 +94,267 @@ pub fn laplacian_smoothing(
     )
 }
 
+/// Performs one iteration of Loop Subdivision on geometry.
+///
+/// The subdivision works in two steps:
+///
+/// 1) Split each triangle into 4 smaller triangles,
+/// 2) Update the position of each vertex of the mesh based on
+///    weighted averages of its neighboring vertex positions,
+///    depending on where the vertex is in the topology and whether
+///    the vertex is newly created, or did already exist.
+///
+/// The geometry **must** be triangulated.
+///
+/// Implementation based on [mdfisher's
+/// post](https://graphics.stanford.edu/~mdfisher/subdivision.html).
+pub fn loop_subdivision(
+    geometry: &Geometry,
+    vertex_to_vertex_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
+    face_to_face_topology: &HashMap<u32, SmallVec<[u32; 8]>>,
+) -> Geometry {
+    #[derive(Debug, Eq)]
+    struct UnorderedPair(u32, u32);
+
+    impl PartialEq for UnorderedPair {
+        fn eq(&self, other: &Self) -> bool {
+            self.0 == other.0 && self.1 == other.1 || self.0 == other.1 && self.1 == other.0
+        }
+    }
+
+    impl Hash for UnorderedPair {
+        fn hash<H: Hasher>(&self, state: &mut H) {
+            cmp::min(self.0, self.1).hash(state);
+            cmp::max(self.0, self.1).hash(state);
+        }
+    }
+
+    let mut vertices: Vec<Point3<f32>> = geometry.vertices().iter().copied().collect();
+
+    // Relocate existing vertices first
+    for (i, vertex) in vertices.iter_mut().enumerate() {
+        let neighbors = &vertex_to_vertex_topology[&cast_u32(i)];
```

Comment:
Is this an equivalent of `&vertex_to_vertex_topology.get(&cast_u32(i))`? Also, you can't be 100% sure the `vertex_to_vertex_topology` belongs to the `geometry` because it comes from outside.

Reply:
> Is this an equivalent of &vertex_to_vertex_topology.get(&cast_u32(i))

Yes, followed by an unwrap.

> Also, you can't be 100% sure the vertex_to_vertex_topology belongs to the geometry because it comes from outside.

Then you get a panic resulting from an incorrect function use - a programmer error. A panic is what you want in this case. We could potentially ignore the vertex if it is not present in the topology, but that would cover up the underlying issue, not a fan of that solution.

I can change one thing though: since the topologies can not be re-used between the iterations (at least not in a simple way), we can just generate them in the function itself and this should be less of an issue (although, there still can be a bug in the topology itself - but once again, we would want that to surface as a panic while we are still developing).

Reply:
sure, let's keep the getter as is. if it properly panics, then it's good.
i'd keep the topology outside: 1. there is a change of overusing an existing one; 2. we'll have it that way in other functions, too

Reply:
> 1. there is a change of overusing an existing one;

Good point, I'll keep it a parameter.